src/tree/updater_quantile_hist.cc

```diff
@@ -88,94 +89,233 @@ bool QuantileHistMaker::UpdatePredictionCache(
-  // use Subtraction Trick
-  for (auto const& node_pair : nodes_for_subtraction_trick_) {
-    hist_.AddHistRow(node_pair.first);
-    SubtractionTrick(hist_[node_pair.first], hist_[node_pair.second],
-                     hist_[(*p_tree)[node_pair.first].Parent()]);
-  }
-  builder_monitor_.Stop("SyncHistograms");
+void QuantileHistMaker::Builder::BuildNodeStat(
+    const GHistIndexMatrix &gmat,
+    DMatrix *p_fmat,
+    RegTree *p_tree,
+    const std::vector<GradientPair> &gpair_h,
+    int32_t nid) {
+
+  // add constraints
+  if (!(*p_tree)[nid].IsLeftChild() && !(*p_tree)[nid].IsRoot()) {
+    auto parent_id = (*p_tree)[nid].Parent();
+    // it's a right child
+    auto left_sibling_id = (*p_tree)[parent_id].LeftChild();
+    auto parent_split_feature_id = snode_[parent_id].best.SplitIndex();
+
+    {
+      spliteval_->AddSplit(parent_id, left_sibling_id, nid, parent_split_feature_id,
+          snode_[left_sibling_id].weight, snode_[nid].weight);
+    }
+  }
 }
 
-void QuantileHistMaker::Builder::BuildLocalHistograms(
-    int *starting_index,
-    int *sync_count,
+void QuantileHistMaker::Builder::BuildNodeStatBatch(
     const GHistIndexMatrix &gmat,
-    const GHistIndexBlockMatrix &gmatb,
+    DMatrix *p_fmat,
     RegTree *p_tree,
-    const std::vector<GradientPair> &gpair_h) {
-  builder_monitor_.Start("BuildLocalHistograms");
-  for (auto const& entry : qexpand_depth_wise_) {
-    int nid = entry.nid;
-    RegTree::Node &node = (*p_tree)[nid];
-    if (rabit::IsDistributed()) {
-      if (node.IsRoot() || node.IsLeftChild()) {
-        hist_.AddHistRow(nid);
-        // in distributed setting, we always calculate from left child or root node
-        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
-        if (!node.IsRoot()) {
-          nodes_for_subtraction_trick_[(*p_tree)[node.Parent()].RightChild()] = nid;
+    const std::vector<GradientPair> &gpair_h,
+    const std::vector<ExpandEntry>& nodes) {
+  perf_monitor.TickStart();
+  for (const auto& node : nodes) {
+    const int32_t nid = node.nid;
+    const int32_t sibling_nid = node.sibling_nid;
+    this->InitNewNode(nid, gmat, gpair_h, *p_fmat, p_tree, &(snode_[nid]), (*p_tree)[nid].Parent());
+    if (sibling_nid > -1) {
+      this->InitNewNode(nid, gmat, gpair_h, *p_fmat, p_tree,
+                        &(snode_[sibling_nid]), (*p_tree)[sibling_nid].Parent());
+    }
+  }
+  for (const auto& node : nodes) {
+    const int32_t nid = node.nid;
+    const int32_t sibling_nid = node.sibling_nid;
+    BuildNodeStat(gmat, p_fmat, p_tree, gpair_h, nid);
+    if (sibling_nid > -1) {
+      BuildNodeStat(gmat, p_fmat, p_tree, gpair_h, sibling_nid);
+    }
+  }
+  perf_monitor.UpdatePerfTimer(TreeGrowingPerfMonitor::timer_name::INIT_NEW_NODE);
+}
+
+template<typename RowIdxType, typename IdxType>
+std::pair<size_t, size_t> PartitionDenseLeftDefaultKernel(const RowIdxType* rid,
```

Comment:
Should we inline this function?

Reply:
I dont see benefits from this, but I added this